### PR TITLE
Fixed issue for two editors and cancelling an image upload

### PR DESCRIPTION
**Pull request by WouterDeKort**
This pull request fixes issues #65 and #66.

#65 is solved by clearing the `uploadField.value `when switching between visible <--> invisible 
#66 is solved by making the `uploadField.id `unique per editor

For #66, I've added a second editor to the test page.

### CODE REVIEW

File: site/index.ts [hunk 1]
```diff
@@ -159,10 +161,23 @@ domReady(() => {
...
-        });
+        };
+
+        const editorInstance = new StacksEditor(place, content.value, options);
+
         // set the instance on the window for developers to poke around in
         // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
         (window as any)["editorInstance"] = editorInstance;
+
+        const secondEditorInstance = new StacksEditor(
```
Comment (hamvocke) [inline site/index.ts:172]: I'm really cool with having a demo page where there are two editor instances since this is not the first bug around multiple editor instances on the same page. I recommend we add the "two editors on the same page" scenario on another site, though. See the `site/variants` directory? That's where we've got a bunch of other demo pages. Let's add the "two editors" scenario there (can just copy the `empty.html` view as a starting point, that should help reduce a lot of clutter). Also let's make sure to link to the new "two editors" page in the "Settings" menu in the top-right corner :)

**Review by hamvocke — CHANGES_REQUESTED** [review]
Verified that the fix works as intended.

Left two comments inline that should hellp make things a little neater.

I wondered if we could add a bunch of tests here and played around a bit. Looks like we can't add unit tests because the `file` input element is really hard to work with in a unit test - but I've got a playwright test coming up that's verifying the entire "close the image uploader" logic, that should help here. I'll add another PR for that test.

File: src/shared/prosemirror-plugins/image-upload.ts [hunk 2]
```diff
@@ -459,6 +459,7 @@ export class ImageUploader implements PluginView {
 
         this.isVisible = isVisible;
         this.image = state?.file || this.image;
+        this.uploadField.value = null;
```
[the remark this answers is not in the document]
Comment (WouterDeKort) [reply src/shared/prosemirror-plugins/image-upload.ts]: I've moved the code to resetUploader as you suggested.

File: site/index.ts (hunk 1 above)
Comment (WouterDeKort) [reply site/index.ts]: I've added a variant named dualeditors, added it to settings and modified index.ts to check if there is a second editor element on the page to avoid null exceptions.

Is it ok how I've done it or did I miss anything?

File: site/variants/dualeditor.html [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    <link rel="stylesheet" href="https://unpkg.com/@stackoverflow/stacks/dist/css/stacks.min.css" />
+    <script src="https://unpkg.com/@stackoverflow/stacks/dist/js/stacks.min.js"></script>
+</head>
+
+<body class="p24">
+    <form class="wmx6 m-auto">
+        <textarea id="content" name="content" class="d-none"></textarea>
+        <h1 class="mt32">Dual Editor Playground</h1>
+        <div class="grid">
+            <div id="example-1" class="mr8"></div>
```
Comment (hamvocke) [inline site/variants/dualeditor.html:19]: Nit, completely up to you: I think it looks nicer if the editors are stacked on top of each other:

```suggestion
        <div class="grid fd-column">
            <div id="example-1" class="mb8"></div>
```

**Review by hamvocke — APPROVED** [review]

File: package-lock.json [hunk 4]
```diff
@@ -1718,7 +1718,6 @@
                 "jest-resolve": "^26.6.2",
                 "jest-util": "^26.6.2",
                 "jest-worker": "^26.6.2",
-                "node-notifier": "^8.0.0",
```
Comment (b-kelly) [inline package-lock.json:1721]: Let's revert the changes to this file. Just curious, what's the output of `node -v` on your machine?

File: site/index.ts [hunk 5]
```diff
@@ -133,7 +135,7 @@ domReady(() => {
 
     // asynchronously load the required bundles
     void import("../src/index").then(function ({ StacksEditor }) {
-        const editorInstance = new StacksEditor(place, content.value, {
+        const options: StacksEditorOptions = {
```
Comment (b-kelly) [inline site/index.ts:138]: Can we revert this change as well? That would allow us to remove the import at the top. Normally I wouldn't care, but I'm very cautious about adding additional imports to this file so I can keep the bundle separate to test lazy loading.

File: site/index.ts [hunk 6]
```diff
@@ -159,10 +161,22 @@ domReady(() => {
...
-        });
+        };
+
+        const editorInstance = new StacksEditor(place, content.value, options);
+
         // set the instance on the window for developers to poke around in
         // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
         (window as any)["editorInstance"] = editorInstance;
+
+        if ( place2 != null)
```
Comment (b-kelly) [inline site/index.ts:172]: In JS, you always want to use `===` and the `!==` equivalent. In this case, we can simply update to:

```suggestion
        if (place2)
```

File: site/index.ts [hunk 7]
```diff
@@ -159,10 +161,22 @@ domReady(() => {
...
+        {
+            const secondEditorInstance = new StacksEditor(place2, content.value, options);
+
+            // set the instance on the window for developers to poke around in
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
+            (window as any)["secondEditorInstance"] = secondEditorInstance;
```
Comment (b-kelly) [inline site/index.ts:178]: This looks fine to me. I've set precedence with the absolute hack above 😂.

No changes necessary.

File: site/index.ts [hunk 8]
```diff
@@ -171,6 +185,13 @@ domReady(() => {
             setDefaultEditor(e.detail.editorType);
         }
     );
+
+    place2?.addEventListener(
```
Comment (b-kelly) [inline site/index.ts:189]: This is fine as-is, but you could even just leave it out if you wanted. All this does is set the default editor on page load to the last mode you used, which isn't super important for our dual editing test.

File: src/shared/prosemirror-plugins/image-upload.ts [hunk 9]
```diff
@@ -359,6 +359,8 @@ export class ImageUploader implements PluginView {
     resetUploader(): void {
         this.resetImagePreview();
         this.hideValidationError();
+
+        this.uploadField.value = null;
```
Comment (b-kelly) [inline src/shared/prosemirror-plugins/image-upload.ts:363]: Nice catch!

No changes needed.

File: src/shared/prosemirror-plugins/image-upload.ts [hunk 10]
```diff
@@ -116,7 +116,7 @@ export class ImageUploader implements PluginView {
         this.uploadField.className = "d-none";
         this.uploadField.accept = "image/*";
         this.uploadField.multiple = false;
-        this.uploadField.id = "fileUpload";
+        this.uploadField.id = "fileUpload" + (Math.random() * 10000).toFixed(0);
```
Comment (b-kelly) [inline src/shared/prosemirror-plugins/image-upload.ts:119]: This looks good for now. We have a concept of an `internalId` on each Editor, but it isn't exposed anywhere or passed down to any of the plugins. I'll think on it, perhaps extracting the method or even potentially adding a linting rule to disallow setting ids to prevent these types of errors in the future.

No changes necessary.

**Review by b-kelly — CHANGES_REQUESTED** [review]
Overall, this looks awesome! Thank you for taking this on. I only have two changes - one is a suggestion and the other is to revert the `package-lock.json file`. Other than that, lgtm!

File: package-lock.json (hunk 4 above)
Comment (WouterDeKort) [reply package-lock.json:1721]: node -v returns v15.14.0. What version are you using?

I ran node i to get all dependencies and that resulted in updating package-lock.json. Can you help me understand why that happened and why I should revert?

Comment (WouterDeKort) [reply package-lock.json]: I reverted the change I made to this file

File: site/index.ts (hunk 5 above)
Comment (WouterDeKort) [reply site/index.ts:138]: But that would mean duplicating the options lines. Just checking if that's what you want? If it's ok, I'll revert the change.

File: src/shared/prosemirror-plugins/image-upload.ts (hunk 10 above)
Comment (hamvocke) [reply src/shared/prosemirror-plugins/image-upload.ts:119]: In this case we need the `id` though. The `Browse` link in the file upload panel is set up as a `label` using the `for` attribute which (as far as I know) has to point to an `id` in order to work. The only alternative would be to nest the upload `input` within the `label` element.

File: site/index.ts (hunk 5 above)
Comment (b-kelly) [reply site/index.ts:138]: Ahhhh, ok that makes sense. It somehow went over my head that you were using it twice. This is perfectly fine then, sorry for the confusion!

File: src/shared/prosemirror-plugins/image-upload.ts (hunk 10 above)
Comment (b-kelly) [reply src/shared/prosemirror-plugins/image-upload.ts:119]: Right, to clarify better, the linter wouldn't so much be disallowing setting ids, but rather disallowing setting ids without some sort of validation/sanitization. For instance, I just added the [eslint-plugin-no-unsanitized](https://github.com/mozilla/eslint-plugin-no-unsanitized) which does this exact type of thing for e.g. `innerHTML`.

**Review by b-kelly — APPROVED** [review]
Nice work! Thanks for making those changes for me. I'll merge this in ASAP.

File: package-lock.json (hunk 4 above)
Comment (b-kelly) [reply package-lock.json]: Mine is reporting `15.5.0`, so not much farther off yours. I'll update and see if the changes come through on my end as well.

As for the "why", generally we don't want to touch package-lock unless we're updating/adding dependencies to package.json. As for why yours updated the package-lock, specifically? I don't know. npm install can be a little finicky at times and I'm not an expert on why. It may be as simple as "your lock file is right and the committed version is wrong" or "new version, new lockfile rules". I'll take a look into it.
